Render clients list using processed DOAP files

wurstsalat3000 opened the pull request:
This adds a new rendering of the clients list to xmpp.org/software/clients.

Client data is gathered by using DOAP files linked in `data/clients.json`.

Clients are sorted by platform and displayed in 'cards'.

The resulting `clients_list_doap.json` is generated automatically by the tooling supplied by this PR. Logos are downloaded (and resized, if necessary) by the tooling as well.

Closes #405 

![image](https://user-images.githubusercontent.com/35428560/167221290-47a47242-1a36-4502-8d2b-6251b984f6f8.png)

Echolon commented:
Hi @linkmauve 

still no luck to find a bit of time? I think this is an awesome feature I would really like to see published.

But I respect if you haven't

horazont commented on tools/prepare_software_list.py:
```diff
@@ -0,0 +1,290 @@
+import shutil
+import sys
+from urllib.parse import urlparse
+from xml.etree import ElementTree as ET
```
https://docs.python.org/3/library/xml.html

> **Warning**
> 
> The XML modules are not secure against erroneous or maliciously constructed data. If you need to parse untrusted or unauthenticated data see the [XML vulnerabilities](https://docs.python.org/3/library/xml.html#xml-vulnerabilities) and [The defusedxml Package](https://docs.python.org/3/library/xml.html#defusedxml-package) sections.

As we are downloading XML from random servers, we have to ensure that it's not going to blow up, especially now that we build the website on our servers again and not in GitHub's or Travis' CI :).

wurstsalat3000 replied:
switched to `defusedxml` for parsing

horazont commented on tools/prepare_software_list.py:
```diff
@@ -0,0 +1,290 @@
+STATIC_PATH = Path('static')
+LOGOS_PATH = STATIC_PATH / 'images' / 'clients'
+
+XSF_CLIENTS_LIST_URL = 'https://raw.githubusercontent.com/xsf/xmpp.org/master/' \
+    'data/clients.json'
```
Shouldn't this be available locally?

wurstsalat3000 replied:
It is indeed, thanks.

horazont commented on tools/prepare_software_list.py:
```diff
@@ -0,0 +1,290 @@
+        print('Error while trying to download from', url)
+        return False
+
+    with open(DOWNLOAD_PATH / path, 'wb') as file_data:
+        file_data.write(file_request.content)
```
Please use [`iter_content`](https://docs.python-requests.org/en/latest/api/#requests.Response.iter_content) and enforce an absolute limit on the file size. We don't want a misconfigured or malicious server fill our disk or RAM (`.content` buffers the entire response in-memory).

A reasonable limit would be 10 MiB IMO.

wurstsalat3000 replied:
Good point, thanks! 10 MiB it is.

horazont commented on tools/prepare_software_list.py:
```diff
@@ -0,0 +1,290 @@
+        file_size = os.path.getsize(file_path)
+    except OSError as error:
+        print('An error occurred while trying to open logo:', error)
+    if file_size > 300000:
```
Nit:

```python
if file_size <= 300000:
    return
```

saves a level of indentation.

horazont commented on tools/prepare_software_list.py:
```diff
@@ -0,0 +1,290 @@
+        print('An error occurred while trying to open logo:', error)
+    if file_size > 300000:
+        try:
+            with Image.open(file_path) as img:
```
I'm not sure I'm happy with PIL processing untrusted data, but I guess that is ok if we ensure that pillow and the backing libraries are up-to-date.

wurstsalat3000 replied:
At least `Pillow` is installed via `pip install --upgrade`, and I didn't pin any version in `requirements.txt`.

wurstsalat3000 replied:
Website builds are now available via Github Actions. This should relax the situation.

horazont commented on tools/prepare_software_list.py:
```diff
@@ -0,0 +1,290 @@
+    if not success:
+        return None
+
+    if extension[1:] in ('png', 'jpg'):
```
What about `jpeg`, `bmp`, ...?

Should we enforce an allowlist on file extensions?

wurstsalat3000 replied:
My intent was to not touch SVG files. I refactored this a bit. But I don't think we need an explicit allowlist, do we?

horazont commented on tools/prepare_software_list.py:
```diff
@@ -0,0 +1,290 @@
+    success = download_file(
+        XSF_CLIENTS_LIST_URL, Path('xsf_clients_list.json'))
+    if not success:
+        sys.exit()
```
[If I read `sys.exit`](https://docs.python.org/3/library/sys.html#sys.exit) correctly, this exits without any message and without error code. Is that truly what you want?

I would've expected an exception raised here, or an error printed and exit with non-zero exit code, because this effectively makes this thing not do what it's supposed to do.

wurstsalat3000 replied:
Since `clients.json` is always available locally, this code isn't necessary anymore.

horazont commented on tools/prepare_software_list.py:
```diff
@@ -0,0 +1,290 @@
+def parse_doap_infos(doap_file: str
+                     ) -> Optional[dict[str, Union[str, list[str], None]]]:
+    '''Parse DOAP file and return infos'''
+    try:
+        doap = ET.parse(
+            DOWNLOAD_PATH / f'doap_files/{doap_file}.doap')
+    except FileNotFoundError:
+        return None
```
This doesn't guard against any parser errors. Any invalid XML of a single individual client would, as I read it, break the entire listing.

Please protect against syntactically invalid XML.

wurstsalat3000 replied:
I'm catching `defusedxml.ElementTree.ParseError` now.

horazont reviewed (changes requested):
Sorry for the brevity, reviewing this in a 10 minute slot before work.

Please address the security concerns before this can get merged, thanks!

wurstsalat3000 commented:
> Sorry for the brevity, reviewing this in a 10 minute slot before work.
> 
> Please address the security concerns before this can get merged, thanks!

@horazont thank you very much for this review! I hope I addressed all points of your feedback?

Echolon commented:
😍😍😍😍😍😍

Neustradamus commented:
@wurstsalat3000: Thanks!
